src/psyclone/psyir/transformations/intrinsics/dotproduct2code_trans.py

```diff
@@ -0,0 +1,321 @@
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: R. W. Ford, STFC Daresbury Lab
+
+'''Module providing a transformation from a PSyIR DOT_PRODUCT operator
+to PSyIR code. This could be useful if the DOT_PRODUCT operator is not
+supported by the back-end or if the performance in the inline code is
+better than the intrinsic.
+
+'''
+from __future__ import absolute_import
+from psyclone.psyir.nodes import BinaryOperation, Assignment, Reference, \
+    Loop, Literal, ArrayReference, Range, Routine
+from psyclone.psyir.symbols import DataSymbol, INTEGER_TYPE, REAL_TYPE, \
+    ArrayType, ScalarType
+from psyclone.psyir.transformations import TransformationError
+from psyclone.psyir.transformations.intrinsics.operator2code_trans import \
+    Operator2CodeTrans
+
+
+def _get_array_bound(vector1, vector2, writer):
+    '''A utility function that returns the appropriate loop bounds (lower,
+    upper and step) for a vector.
+    If either of the vectors are declared with known bounds (an integer or a
+    symbol) then these bound values are used. If the size is unknown
+    (a deferred or attribute type) then the LBOUND and UBOUND PSyIR
+    nodes are used.
+
+    :param array: the reference that we are interested in.
+    :type array: :py:class:`psyir.nodes.Reference`
+    :param int index: the (array) reference index that we are \
+        interested in.
+
+    :returns: the loop bounds for this array index.
+    :rtype: (Literal, Literal, Literal) or \
+        (BinaryOperation, BinaryOperation, Literal)
+
+    :raises TransformationError: if both vector1 and vector2 are not \
+        references to array symbols.
+
+    '''
+    if not ((isinstance(vector1, Reference) and vector1.symbol.is_array) or
+            (isinstance(vector2, Reference) and vector2.symbol.is_array)):
+        raise TransformationError(
+            f"DotProduct2CodeTrans._get_array_bound requires at least one of "
+            f"the dotproduct arguments to be an array but found "
+            f"'{writer(vector1)}' and '{writer(vector2)}'.")
+
+    # Look for explicit bounds in one of the array declarations
+    for vector in [vector1, vector2]:
+        symbol = vector.symbol
+        my_dim = symbol.shape[0]
```

Comment:
What about `dot_product(a(1,:),b(2,:))`?

Reply:
Again, you could raise a NotImplementedError for the time being given the time pressures.

Reply:
The validate method ensures this as there is a constraint that array slices must be for the first dimension of an array.